File: src/test/java/nz/co/searchwellington/filters/attributesetters/PublisherPageAttributeSetterTest.java
package nz.co.searchwellington.filters.attributesetters;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.when;
import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.repositories.HibernateResourceDAO;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.mock.web.MockHttpServletRequest;

public class PublisherPageAttributeSetterTest {
	
	@Mock HibernateResourceDAO resourceDAO;
	@Mock Website publisher;
	private MockHttpServletRequest request;
	private PublisherPageAttributeSetter pageAttributeSetter;
	
	@Before
	public void setup() {
		MockitoAnnotations.initMocks(this);
		request = new MockHttpServletRequest();
		when(publisher.getName()).thenReturn("Wellington City Council");
		when(resourceDAO.getPublisherByUrlWords("wellington-city-council")).thenReturn(publisher);
		pageAttributeSetter = new PublisherPageAttributeSetter(resourceDAO);		
	}
	
	@Test
	public void shouldSetPublisherAttributeForPublisherPath() {
		request.setPathInfo("/wellington-city-council");

		pageAttributeSetter.setAttributes(request);
		
		assertEquals(publisher, request.getAttribute("publisher"));		
	}
	
	@Test
	public void shouldSetPublisherAttributeForPublisherEditPath() {
		request.setPathInfo("/wellington-city-council/edit");
		
		pageAttributeSetter.setAttributes(request);
		
		assertEquals(publisher, request.getAttribute("publisher"));		
	}
	
	@Test
	public void shouldSetPublisherAttributeForPublisherRssPath() {
		request.setPathInfo("/wellington-city-council/rss");
		
		pageAttributeSetter.setAttributes(request);
		
		assertEquals(publisher, request.getAttribute("publisher"));		
	}
	
}

File: src/test/scala/nz/co/searchwellington/filters/attributesetters/PublisherPageAttributeSetterTest.scala
package nz.co.searchwellington.filters.attributesetters

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.Website
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.mockito.Mockito.{mock, when}
import org.springframework.mock.web.MockHttpServletRequest

import java.util.UUID
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}

class PublisherPageAttributeSetterTest extends ReasonableWaits {
  private val mongoRepository: MongoRepository = mock(classOf[MongoRepository])
  private val publisher = Website(id = UUID.randomUUID().toString, title = "Wellington City Council")
  private val request = new MockHttpServletRequest

  private val publisherPageAttributeSetter = new PublisherPageAttributeSetter(mongoRepository)

  @BeforeEach
  def setup(): Unit = {
    when(mongoRepository.getWebsiteByUrlwords("wellington-city-council")).thenReturn(Future.successful(Some(publisher)))
  }

  @Test
  def shouldSetPublisherAttributeForPublisherPath(): Unit = {
    request.setRequestURI("/wellington-city-council")

    val attributes = Await.result(publisherPageAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(publisher, attributes("publisher"))
  }


  @Test
  def shouldSetPublisherAttributeForPublisherArchivePath(): Unit = {
    request.setRequestURI("/wellington-city-council/2020-mar")

    val attributes = Await.result(publisherPageAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(publisher, attributes("publisher"))
  }

  @Test
  def shouldSetPublisherAttributeForPublisherRssPath(): Unit = {
    request.setRequestURI("/wellington-city-council/rss")

    val attributes = Await.result(publisherPageAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(publisher, attributes("publisher"))
  }

}
